### PR TITLE
Create gridpack::environment with existing MPI communicator

### PR DESCRIPTION
**Pull request by wperkins**
This branch adds a facility to create a `gridpack::environment` with an existing MPI communicator.  This allows an external application to call `MPI_Initialize()` and instantiate multiple GridPACK environments on an externally distribution of MPI ranks.  
**NOTE**: This GridPACK branch depends on a specific branch of [Global Arrays](https://github.com/GlobalArrays/ga/tree/develop) (`feature/init-comm`).  This PR should not be accepted until the necessary changes have been merged into the GA `master` branch (and the GridPACK GA submodule is updated and actually works).

### CODE REVIEW

**Comment by abhyshr** [comment]
1. This branch needs to resolve the conflicts first.
2. Don't understand what is the additional functionality being added here. Is the ability to set an MPI_Comm with GridPACK's environment object & being able to set it via Python?

**Comment by wperkins** [comment]
> 2\. Don't understand what is the additional functionality being added here. Is the ability to set an MPI_Comm with GridPACK's environment object & being able to set it via Python?

GridPACK expects to control MPI, that is GridPACK calls `MPI_Init()` and `MPI_Finalize()` to set up an `environment`.  The changes in this PR simply allows a GridPACK `environment` to be created with an externally created `MPI_Comm`.  And, not just in Python. 

This was set up for the [ExaLearn](https://github.com/exalearn) folks, who drive multiple GridPACK instances with an MPI-based  driver.

**Comment by wperkins** [comment]
> 1. This branch needs to resolve the conflicts first.

Agreed.  It should be noted, however, that this will require reconciling the HADREC application with current dynamic simulation API without alienating the [ExaLearn](https://github.com/exalearn) application of HADREC.  I do not know details.

**Comment by wperkins** [comment]
Note to self: The value of the CMake variable `ENABLE_ENVIRONMENT_FROM_COMM` needs to be in `GridPACK.cmake` and the Python interface adjusted to allow `ENABLE_ENVIRONMENT_FROM_COMM=FALSE`.

File: src/environment/environment.cpp [hunk 1]
```diff
@@ -83,22 +86,65 @@ Environment::Environment(int argc, char **argv,const char* help,const long int&
...
+  }
+  MPI_Abort(comm, 1);
+#endif
+}
+
 Environment::~Environment(void)
 {
-  // Finalize math libraries
-  gridpack::math::Finalize();
+  if (p_from_comm) {
```
Comment (wperkins) [inline src/environment/environment.cpp]: Note the artifacts from rebase.
I don't understand the destructor.  It seems to me that `gridpack::math::Finalize();` and `GA_Terminate()` should be called no matter the state of `p_from_comm`.

File: src/applications/dynamic_simulation_full_y/CMakeLists.txt [hunk 2]
```diff
@@ -198,4 +215,11 @@ install(TARGETS dsf2.x DESTINATION bin)
 # run application as test
 # -------------------------------------------------------------
 gridpack_add_run_test("dynamic_simulation_full_y" dsf.x input_145.xml)
+<<<<<<< HEAD
```
Comment (wperkins) [inline src/applications/dynamic_simulation_full_y/CMakeLists.txt]: Rebase artifacts.

**Comment by bjpalmer** [comment]
I'm in the process of cleaning up after the rebase. I'm not sure how I managed to overlook some of the artifacts since the code doesn't compile. I noticed the same thing about the destructor and fixed it.

File: src/applications/dynamic_simulation_full_y/dsf_comm_main.cpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+      gridpack::dynamic_simulation::DSFullBranch>(ds_network);
+
+    // transfer results from PF calculation to DS calculation
+    transferPFtoDS(pf_network, ds_network); 
+
+    // read in faults from input file
+    //gridpack::utility::Configuration::CursorPtr cursor;
+    cursor = config->getCursor("Configuration.Dynamic_simulation");
+    std::vector<gridpack::dynamic_simulation::Event> faults;
+    faults = ds_app.getFaults(cursor);
```
Comment (wperkins) [inline src/applications/dynamic_simulation_full_y/dsf_comm_main.cpp]: Also, `getFaults()` is no longer a method.  I think it's `getEvents()` now. I don't know if there is an internal difference.

Comment (bjpalmer) [reply src/applications/dynamic_simulation_full_y/dsf_comm_main.cpp]: FIxed.

**Comment by abhyshr** [comment]
@bjpalmer: any updates on when GA will be updated to have fewture/init-comm in develop?

**Comment by bjpalmer** [comment]
The stuff in feature/init-comm should already be in the develop branch of GA. I've been linking GridPACK against the develop branch and it seems to work okay.

**Comment by wperkins** [comment]
> The stuff in feature/init-comm should already be in the develop branch of GA. I've been linking GridPACK against the develop branch and it seems to work okay.

FYI. In the GridPACK `feature/init-comm` branch, if you specify `-D BUILD_GA=YES` the GA `develop` branch will be built.

**Comment by abhyshr** [comment]
So then can we just use the develop branch of GA? And merge GridPack's feature/init-comm to develop? I can update the install script to use GA's develop branch. Thoughts?

**Comment by abhyshr** [comment]
First we need to resolve the issue with GridPACK python application hanging on multiple processors. @bjpalmer: do you have time to take a look at this?

**Comment by bjpalmer** [comment]
You should be able to switch the install script to use develop. Is the code hanging when using the progress ranks runtime in GA? You need to be careful with that runtime to get the code to exit cleanly.

**Comment by abhyshr** [comment]
I am not using progress ranks. You can try the updated 39 bus python example I've pushed to the branch

**Comment by wperkins** [comment]
> You should be able to switch the install script to use develop. Is the code hanging when using the progress ranks runtime in GA? You need to be careful with that runtime to get the code to exit cleanly.

You can still use the official GA release to build this branch. But, if you do, you cannot specify `-D ENABLE_ENVIRONMENT_FROM_COMM=TRUE`. It would be nice to rework `FindGA.cmake` to test for the necessary GA routine and set that option automatically. That's a lot of work though. 

Personally, I think we should merge this. The audience for `-D ENABLE_ENVIRONMENT_FROM_COMM=TRUE` is small and they are already navigating it.

**Comment by wperkins** [comment]
@bjpalmer, I've added some very brief documentation.  If it's enough, please approve this PR and I'll merge it.  Thanks.

**Comment by abhyshr** [comment]
Good to merge after rebase.

**Review by bjpalmer — APPROVED** [review]

**Pull request merged**